agent/agent-tooling/src/main/java/com/microsoft/applicationinsights/agent/internal/exporter/LoggerExporter.java

```diff
@@ -0,0 +1,359 @@
+
+  private static final AttributeKey<String> AI_OPERATION_NAME_KEY =
+      AttributeKey.stringKey("applicationinsights.internal.operation_name");
+
+  private static final OperationLogger exportingLogLogger =
+      new OperationLogger(Exporter.class, "Exporting log");
+
+  private final TelemetryClient telemetryClient;
+  private final AtomicBoolean stopped = new AtomicBoolean();
+
+  public LoggerExporter(TelemetryClient telemetryClient) {
+    this.telemetryClient = telemetryClient;
+  }
+
+  @Override
+  public CompletableResultCode export(Collection<LogData> logs) {
+    if (CoreUtils.isNullOrEmpty(TelemetryClient.getActive().getInstrumentationKey())) {
+      logger.debug("Instrumentation key is null or empty. Fail to export logs.");
+      return CompletableResultCode.ofFailure();
+    }
+
+    if (stopped.get()) {
+      logger.debug("LoggerExporter has been stopped. Fail to export logs.");
+      return CompletableResultCode.ofFailure();
+    }
+
+    boolean failure = false;
+    for (LogData log : logs) {
+      try {
+        internalExport(log);
+        exportingLogLogger.recordSuccess();
+      } catch (Throwable t) {
+        exportingLogLogger.recordFailure(t.getMessage(), t);
+        failure = true;
+      }
+    }
+
+    return failure ? CompletableResultCode.ofFailure() : CompletableResultCode.ofSuccess();
+  }
+
+  @Override
+  public CompletableResultCode flush() {
+    return CompletableResultCode.ofSuccess();
+  }
+
+  @Override
+  public CompletableResultCode shutdown() {
+    stopped.set(true);
+    return CompletableResultCode.ofSuccess();
+  }
+
+  private void internalExport(LogData log) {
+    int severity = log.getSeverity().getSeverityNumber();
+    int threshold = getThreshold().getSeverityNumber();
+    if (severity < threshold) {
+      return;
+    }
+
+    logger.debug("exporting log: {}", log); // do we need to log this, will that be too much?
+    String stack = log.getAttributes().get(SemanticAttributes.EXCEPTION_STACKTRACE);
+    if (stack == null) {
+      trackMessage(log);
+    } else {
+      trackMessageAsException(log, stack);
+    }
+  }
+
+  private void trackMessage(LogData log) {
+    MessageTelemetryBuilder telemetryBuilder = telemetryClient.newMessageTelemetryBuilder();
+
+    Attributes attributes = log.getAttributes();
+
+    // set standard properties
+    telemetryBuilder.setTime(FormattedTime.offSetDateTimeFromEpochNanos(log.getEpochNanos()));
+    setOperationTags(telemetryBuilder, log);
+    setSampleRate(telemetryBuilder, log);
+    setExtraAttributes(telemetryBuilder, attributes);
+
+    telemetryBuilder.setSeverityLevel(toSeverityLevel(log.getSeverity()));
+    telemetryBuilder.setMessage(log.getBody().asString());
+
+    // set message-specific properties
+    setLoggerProperties(
+        telemetryBuilder,
+        log.getInstrumentationLibraryInfo().getName(),
+        attributes.get(SemanticAttributes.THREAD_NAME),
+        log.getSeverity());
```

Comment:
What's the difference between `telemetryBuilder.addProperty("LoggingLevel", mapSeverityToLoggingLevel(severity).toString());` and `telemetryBuilder.setSeverityLevel(toSeverityLevel(log.getSeverity()));`?

Reply:
they're two different properties. LoggingLevel is something we sent in the cusomDimensions and severity is a property required by the common schema.